Assignment4/z170395-Assigment4/task07.py

```diff
@@ -0,0 +1,207 @@
+    if str(o) == str(living_thing):
+        subclasses.add(s)
+        #print(s,p,o)
+
+# Imprimir las subclases encontradas
+for subclass in subclasses:
+    print(subclass)
+
+"""**TASK 7.2: List all individuals of "Person" with RDFLib and SPARQL (remember the subClasses)**
+
+"""
+
+# TO DO
+#SPARQL
+person = Namespace("http://somewhere#Person")
+
+# Define a SPARQL query to select all individuals of "Person" (including subclasses)
+query = prepareQuery(
+    f"""
+    SELECT ?individual
+    WHERE {{
+        ?individual a/rdfs:subClassOf* <{person}> .
+    }}
+    """
+)
+
+# Execute the SPARQL query and print the results
+for row in g.query(query):
+    print(row["individual"])
+
+# Visualize the results
+
+rdf = Namespace("http://www.w3.org/1999/02/22-rdf-syntax-ns#")
+rdfs = Namespace("http://www.w3.org/2000/01/rdf-schema#")
+person = Namespace("http://somewhere#Person")
+
+individuals = set()
+
+# Iterar a través de todas las tripletas y encontrar instancias de "Person" y sus subclases
+subclasses = set()
+
+# Obtener todas las tripletas en el grafo donde el objeto es una subclase de "Person"
+for s, p, o in g.triples((None, rdfs.subClassOf, None)):
+  if str(o) == person:
+    subclasses.add(s)
+
+for s, p, o in g.triples((None, rdf.type, None)):
+    #print(s,p,o)
+    if str(o) == str(person) or str(o) in str(subclasses):
+        individuals.add(s)
+
+# Imprimir las instancias encontradas
+for individual in individuals:
+    print(individual)
+
+
+
+"""**TASK 7.3: List all individuals of "Person" or "Animal" and all their properties including their class with RDFLib and SPARQL. You do not need to list the individuals of the subclasses of person**
+
+"""
+
+# TO DO
+#SPARQL
+
+person = Namespace("http://somewhere#Person")
+animal = Namespace("http://somewhere#Animal")
+
+# Definir una consulta SPARQL para seleccionar todas las instancias de "Person" sin incluir sus subclases
+query = prepareQuery(
+f"""
+PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>
+PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>
+SELECT ?individual
+    WHERE {{
+        {{
+            ?individual rdf:type <{person}> .
+        }}
+        UNION
+        {{
+            ?individual rdf:type <{animal}> .
+        }}
+    }}
+    """
+)
+
+# Ejecutar la consulta SPARQL e imprimir los resultados
+for row in g.query(query):
+    print(row["individual"])
+
+
+# Visualize the results
+
+individuals = set()
+for s, p, o in g.triples((None, RDF.type, None)):
+    if str(o) == person or str(o) == animal:
+        individuals.add(s)
+for individual in individuals:
+    print(individual)
+
+"""**TASK 7.4:  List the name of the persons who know Rocky**"""
+
+# TO DO
+knows = "http://xmlns.com/foaf/0.1/knows"
+rocky = "http://somewhere#RockySmith"
+fn = "http://www.w3.org/2001/vcard-rdf/3.0/FN"
+individuals = set()
+names = set()
+for s, p, o in g.triples((None,None, None )):
+  # print(s,p,o)
+   if str(p) == knows and str(o) == rocky :
```

Comment:
You have not tested this query. None of the properties are `knows`. They are URIs